actors/builtin/miner/deadline_state.go

```diff
@@ -444,3 +446,161 @@ func (dl *Deadline) popExpiredPartitions(store adt.Store, until abi.ChainEpoch,
 
 	return popped, modified, nil
 }
+
+// RemovePartitions removes the specified partitions, shifting the remaining
+// ones to the left, and returning the live and dead sectors they contained.
+//
+// Returns an error if any of the partitions contained faulty sectors or early
+// terminations.
+func (dl *Deadline) RemovePartitions(store adt.Store, partitions *bitfield.BitField, quant QuantSpec) (
+	live, dead *abi.BitField, removedPower PowerPair, err error,
+) {
+	oldPartitions, err := dl.PartitionsArray(store)
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to load partitions: %w", err)
+	}
+
+	partitionCount := oldPartitions.Length()
+	toCompact, err := partitions.AllMap(partitionCount)
+	if err != nil {
+		// TODO: This may be an illegal argument error if we have too many partitions.
+		// https://github.com/filecoin-project/specs-actors/issues/597
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to expand partitions into map: %w", err)
+	}
+
+	// Nothing to do.
+	if len(toCompact) == 0 {
+		return bitfield.NewFromSet(nil), bitfield.NewFromSet(nil), NewPowerPairZero(), nil
+	}
+
+	for partIdx := range toCompact { //nolint:nomaprange
+		if partIdx >= partitionCount {
+			// TODO: This is an illegal argument error
+			// https://github.com/filecoin-project/specs-actors/issues/597
+			return nil, nil, NewPowerPairZero(), xerrors.Errorf("partition index %d out of range [0, %d)", partIdx, partitionCount)
+		}
+	}
+
+	// Should already be checked earlier, but we might as well check again.
+	noEarlyTerminations, err := dl.EarlyTerminations.IsEmpty()
+	if err != nil {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("failed to check for early terminations: %w", err)
+	}
+	if !noEarlyTerminations {
+		return nil, nil, NewPowerPairZero(), xerrors.Errorf("cannot compact deadline with early terminations: %w", err)
+	}
+
+	newPartitions := adt.MakeEmptyArray(store)
+	allDeadSectors := make([]*bitfield.BitField, 0, len(toCompact))
+	allLiveSectors := make([]*bitfield.BitField, 0, len(toCompact))
+	removedPower = NewPowerPairZero()
+
+	// Define all of these out here to save allocations.
+	var (
+		lazyPartition cbg.Deferred
+		byteReader    bytes.Reader
+		partition     Partition
+	)
+	if err = oldPartitions.ForEach(&lazyPartition, func(partIdx int64) error {
+		// If we're keeping the partition as-is, append it to the new partitions array.
+		if _, ok := toCompact[uint64(partIdx)]; !ok {
+			return newPartitions.AppendContinuous(&lazyPartition)
+		}
+
+		// Ok, actually unmarshal the partition.
+		byteReader.Reset(lazyPartition.Raw)
+		err := partition.UnmarshalCBOR(&byteReader)
+		byteReader.Reset(nil)
+		if err != nil {
```

Comment:
Nit: fold into line 512?

Reply:
I wanted to reset the byte reader in all cases. It's not really an issue here, it just makes me feel a bit safer.